File: Pythoncode/62.不同路径.py
class Solution:
    def countPaths(self, index2NumofPaths, m, n, x, y):
        if (x, y) in index2NumofPaths.keys():
            return index2NumofPaths[(x, y)]
        if x == 0 and y == 0:
            index2NumofPaths[(x, y)] = 1
            return 1
        elif x < 0 or x >= m or y < 0 or y >= n:
            index2NumofPaths[(x, y)] = 0
            return 0
        else:
            index2NumofPaths[(x, y)] = self.countPaths(index2NumofPaths, m, n, x-1, y) + self.countPaths(index2NumofPaths, m, n, x, y-1)
            return index2NumofPaths[(x, y)]

    def uniquePaths(self, m: int, n: int) -> int:
        index2NumofPaths = {}
        return self.countPaths(index2NumofPaths, m, n, m-1, n-1)

s = Solution()
print(s.uniquePaths(23, 12))

File: Cppcode/62.不同路径.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int uniquePaths(int m, int n) {
        vector<vector<int>> dp(m, vector<int>(n));
        for (auto i = 0; i < m; i++)
        {
            for (auto j = 0; j < n; j++)
            {
                if (i == 0 && j == 0)
                {
                    dp[i][j] = 1;
                    continue;
                }
                auto a = i-1>=0 ? dp[i-1][j] : 0;
                auto b = j-1>=0 ? dp[i][j-1] : 0;
                dp[i][j] = a + b;
            }
        }
        return dp[m-1][n-1];
    }
};